File: ruby/112.path-sum.rb
class TreeNode
  attr_accessor :val, :left, :right

  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

# @param {TreeNode} root
# @param {Integer} target_sum
# @return {Boolean}
def has_path_sum(root, target_sum)
  dfs?(root, target_sum)
end

def dfs?(node, target_sum)
  return false if node.nil?
  return target_sum == node.val if node.left.nil? && node.right.nil?

  dfs?(node.left, target_sum - node.val) || dfs?(node.right, target_sum - node.val)
end

File: rust/src/112.path-sum.rs
use crate::Solution;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

#[allow(dead_code)]
impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

use std::cell::RefCell;
use std::rc::Rc;

#[allow(dead_code)]
impl Solution {
    pub fn has_path_sum(root: Option<Rc<RefCell<TreeNode>>>, target_sum: i32) -> bool {
        Self::dfs(root, target_sum)
    }

    fn dfs(root: Option<Rc<RefCell<TreeNode>>>, target_sum: i32) -> bool {
        match root {
            None => false,
            Some(node) => {
                let node = node.borrow();

                if node.left.is_none() && node.right.is_none() {
                    return target_sum - node.val == 0;
                }

                Self::dfs(node.left.clone(), target_sum - node.val)
                    || Self::dfs(node.right.clone(), target_sum - node.val)
            }
        }
    }
}
